### PR TITLE
Feat(condition): Check if attribute is present in player fleet

### PR DESCRIPTION
**Pull request by RisingLeaf**
as needed by #10139

## Summary
Creates a new condition `"ship attribute: <attribute>"` that returns the sum of that attribute across the whole fleet
## Testing Done
yes, works

### CODE REVIEW

File: source/PlayerInfo.cpp [hunk 1]
```diff
@@ -3475,6 +3475,20 @@ void PlayerInfo::RegisterDerivedConditions()
 		return retVal;
 	});
 
+	// Check if an attribute is present in fleet.
+	auto &&shipAttributeProvider = conditions.GetProviderPrefixed("ship attribute: ");
+	auto shipAttributeFun = [this, flagshipAttributeHelper](const string &name) -> int64_t
```
Comment (RisingLeaf) [inline source/PlayerInfo.cpp:3480]:
```suggestion
	auto shipAttributeFun = [this](const string &name) -> int64_t
```

**Comment by RisingLeaf** [comment]
tested, works

**Comment by tibetiroka** [comment]
I think it could be confusing to call it `ship attribute` when it is checking the fleet. Maybe it should be `fleet attribute`?

**Comment by tibetiroka** [comment]
Would it make sense to have it return the sum of the attribute amount, instead of just a boolean? It could still be used a a boolean, but would enable more functionality.

**Comment by RisingLeaf** [comment]
> Would it make sense to have it return the sum of the attribute amount, instead of just a boolean? It could still be used a a boolean, but would enable more functionality.

sure, easy enough

**Comment by RisingLeaf** [comment]
done

File: source/PlayerInfo.cpp [hunk 2]
```diff
@@ -3475,6 +3475,20 @@ void PlayerInfo::RegisterDerivedConditions()
 		return retVal;
 	});
 
+	// Check if an attribute is present in fleet.
```
Comment (tibetiroka) [inline source/PlayerInfo.cpp:3478]:
```suggestion
	// Check if an attribute is present in the player's fleet.
```

File: source/PlayerInfo.cpp [hunk 3]
```diff
@@ -3475,6 +3475,20 @@ void PlayerInfo::RegisterDerivedConditions()
...
+	auto &&shipAttributeProvider = conditions.GetProviderPrefixed("ship attribute: ");
+	auto shipAttributeFun = [this](const string &name) -> int64_t
+	{
+		auto attrib = name.substr(strlen("ship attribute: "));
+		int64_t sum = 0;
+		for(const shared_ptr<Ship> &ship : ships)
+			sum += ship->Attributes().Get(attrib);
+
```
Comment (tibetiroka) [inline source/PlayerInfo.cpp:3486]:
```suggestion
```
This is more in line with the functions above. I'm wondering if the variable should be renamed to `retVal` for consistency as well, I'm not sure.

File: source/PlayerInfo.cpp [hunk 4]
```diff
@@ -3475,6 +3475,19 @@ void PlayerInfo::RegisterDerivedConditions()
...
 
+	// Check if an attribute is present in the player's fleet.
+	auto &&shipAttributeProvider = conditions.GetProviderPrefixed("ship attribute: ");
+	auto shipAttributeFun = [this](const string &name) -> int64_t
+	{
+		auto attrib = name.substr(strlen("ship attribute: "));
+		int64_t sum = 0;
+		for(const shared_ptr<Ship> &ship : ships)
+			sum += ship->Attributes().Get(attrib);
+		return sum;
```
Comment (RisingLeaf) [inline source/PlayerInfo.cpp:3486]:
```suggestion
		int64_t retVal = 0;
		for(const shared_ptr<Ship> &ship : ships)
			retVal += ship->Attributes().Get(attrib);
		return retVal;
```

File: source/PlayerInfo.cpp [hunk 5]
```diff
@@ -3475,6 +3475,19 @@ void PlayerInfo::RegisterDerivedConditions()
 		return retVal;
 	});
 
+	// Check if an attribute is present in the player's fleet.
+	auto &&shipAttributeProvider = conditions.GetProviderPrefixed("ship attribute: ");
+	auto shipAttributeFun = [this](const string &name) -> int64_t
+	{
+		auto attrib = name.substr(strlen("ship attribute: "));
```
Comment (TomGoodIdea) [inline source/PlayerInfo.cpp:3482]: death to auto!
```suggestion
		string attrib = name.substr(strlen("ship attribute: "));
```

Comment (RisingLeaf) [reply source/PlayerInfo.cpp:3482]: meh

**Review by TomGoodIdea — APPROVED** [review]

**Comment by ziproot** [comment]
Should there be a distinction between “present and active” and “all ships” like there is for “ships” or “outfit?”

The version of this for the flagship multiples the value by 1000 due to the potential for decimal values. Will that be done here?

Where is the associated wiki description PR?

Thanks.

**Comment by RisingLeaf** [comment]
> Should there be a distinction between “present and active” and “all ships” like there is for “ships” or “outfit?”

maybe, not in this PR
 
> The version of this for the flagship multiples the value by 1000 due to the potential for decimal values. Will that be done here?

yes

> Where is the associated wiki description PR?

I was not aware that is a necessity? The devs wont add that stuff anymore?

> Thanks.

you are welcome

**Comment by ziproot** [comment]
> I was not aware that is a necessity? The devs wont add that stuff anymore?

It is now, as of https://github.com/endless-sky/endless-sky/commit/314dc21f9839d3feedd551830a4ea338d94dc7f8, because anyone can make a wiki PR now.

**Comment by Amazinite** [comment]
I'll handle wiki updates for any PRs that I merge that were opened prior to the updated template, but going forward it will be the responsibility of the PR opener to also open a PR to the wiki repository when necessary, just as they would for the high DPI or assets repositories.

File: source/PlayerInfo.cpp [hunk 6]
```diff
@@ -3477,6 +3477,19 @@ void PlayerInfo::RegisterDerivedConditions()
...
 	});
 
+	// Check if an attribute is present in the player's fleet.
+	auto &&shipAttributeProvider = conditions.GetProviderPrefixed("ship attribute: ");
+	auto shipAttributeFun = [this](const string &name) -> int64_t
+	{
+		string attrib = name.substr(strlen("ship attribute: "));
+		int64_t retVal = 0;
+		for(const shared_ptr<Ship> &ship : ships)
+			retVal += ship->Attributes().Get(attrib) * 1000.;
```
Comment (Amazinite) [inline source/PlayerInfo.cpp:3487]: This should be making use of the `flagshipAttributeHelper` function (which could be renamed to just `shipAttributeHelper`). I also think that the search space of this condition should match the search space of the other "default" conditions that check your ships. That is, `"ship attribute: <attr>"` should only look for ships which are not parked or disabled and in your system, whereas `"ship attribute (all): <attr>"` would check everything that isn't destroyed. (See `"ships: <category>"` and `"ship model: <model>"` just above this for examples.)

**Comment by tibetiroka** [comment]
@RisingLeaf Can you get around to resolving the review comments any time soon?

**Comment by tibetiroka** [comment]
Closing as orphaned. Feel free to reopen this PR once the review comments are addressed.

**Comment by TomGoodIdea** [comment]
#11306

**Comment by Amazinite** [comment]
I forgot this even existed.

**Pull request closed without merging**